### ISSUE TITLE
[api] - a issue tracker delayed job fails

### ISSUE THREAD

**Issue by mmohring** [openSUSE/open-build-service#3405]
Issue/Feature description
====================
I use the current git master OBS 772e399278c808f49bad9287645d8faaf1c8fa51.
In the production.log file, I see:

E, [2017-07-16T21:51:33.604996 #7916] ERROR -- : [7916:4657.44] 2017-07-16T21:51:33+0200: [Worker(delayed_job.1010 host:mercury pid:7916)] Job Delayed::PerformableMethod (id=1199179) (queue=issuetracking) FAILED permanently with Delayed::DeserializationError: Job failed to load: undefined class/module ActiveModel::Type::Text. Handler: "--- !ruby/object:Delayed::PerformableMethod\nobject: !ruby/object:IssueTracker\n  raw_attributes:\n .............

Expected result
============
The issue trackers delayed job should not fail.

How to Reproduce
===============
enable the CVE tracker, as it is dumped in:

$ osc -A https://api.opensuse.org api issue_trackers > issue-trackers-cve.xml
select the section for CVE tracker from the dump and save to issue-trackers-cve.xml
activate the tracker with:
$ osc -A MYOBSURL api -m POST -T issue-trackers-cve.xml issue_trackers

Further information
===============
If you tell me what you need more, I will provide this information.

**Comment by mmohring** [comment]
Calling:
$ osc -A https://api.osadl.org api -m DELETE issue_trackers/cve
hangs up

**Comment by mmohring** [comment]
causes in the logfile:

I, [2017-07-16T22:56:50.726008 #19890]  INFO -- : [93c1638d-d6a6-4320-bc10-aba46d4e198e] [19890:7636.91] Completed 500 Internal Server Error in 75673ms (ActiveRecord: 52397.5ms | Backend: 0.0ms | XML: 0.0ms)
F, [2017-07-16T22:56:50.750883 #19890] FATAL -- : [93c1638d-d6a6-4320-bc10-aba46d4e198e] [19890:7636.93]   
F, [2017-07-16T22:56:50.751102 #19890] FATAL -- : [93c1638d-d6a6-4320-bc10-aba46d4e198e] [19890:7636.93] ActiveRecord::StatementInvalid (Mysql2::Error: Lock wait timeout exceeded; try restarting transaction: DELETE FROM `package_issues` WHERE `package_issues`.
`issue_id` = 2165):
F, [2017-07-16T22:56:50.751164 #19890] FATAL -- : [93c1638d-d6a6-4320-bc10-aba46d4e198e] [19890:7636.93]   
F, [2017-07-16T22:56:50.751229 #19890] FATAL -- : [93c1638d-d6a6-4320-bc10-aba46d4e198e] [19890:7636.93] app/controllers/issue_trackers_controller.rb:115:in `destroy'

**Comment by mmohring** [comment]
so let me conclude:
1. The issue trackers delayed jobs fail
2. You cannot delete a issue tracker
3. An update works in the API but still lets the delayed jobs fail
4. A disabled issue tracker cannot be enabled by the API

**Comment by bgeuken** [comment]
Did you try to restart delayed jobs?

**Comment by mmohring** [comment]
Of course. When calling the DELETE I also tried without delayed jobs running to not utilize MySQL that much while it tries to grab the issue database. No success. My idea was if I could get away the delayed job error by deleting the issue tracker entry for CVE.

**Comment by mmohring** [comment]
It looks like I can also not change the enable flag of a tracker. The next Idea I had was to disable the CVE tracker by setting:
`<enable-fetch>false</enable-fetch>`
in the xml file for a tracker before calling :
$ osc -A MYAPIURL api -m POST -T issue-trackers-cve.xml issue_trackers
But subsequent get shows tracker is still enabled:
$ osc -A MYAPIURL api issue_trackers/cve | grep enable-
`<enable-fetch>true</enable-fetch>`

**Comment by bgeuken** [comment]
Thank you for clarifying.

I am currently not able to reproduce this. I would like to verify that delayed jobs daemon didn't get stuck on your machine. Could you stop delayed jobs and then grep for "delayed_job"?

And out of curiosity, if you stop delayed jobs and then run the delete request; does it work?

**Comment by evanrolfe** [comment]
> In the production.log file, I see:

E, [2017-07-16T21:51:33.604996 #7916] ERROR -- : [7916:4657.44] 2017-07-16T21:51:33+0200: [Worker(delayed_job.1010 host:mercury pid:7916)] Job Delayed::PerformableMethod (id=1199179) (queue=issuetracking) FAILED permanently with Delayed::DeserializationError: Job failed to load: undefined class/module ActiveModel::Type::Text. Handler: "--- !ruby/object:Delayed::PerformableMethod\nobject: !ruby/object:IssueTracker\n raw_attributes:\n .............

@mmohring can you please give us the full line of that log entry?

**Comment by mmohring** [comment]
Sure. I have in the meantime tested on a separate test instance for OBS. There I can delete an issue tracker, and add one. Behavior:

1. If a tracker has  `<enable-fetch>false</enable-fetch>` set, you cant overwrite it with `<enable-fetch>true</enable-fetch>`. You have to delete the tracker, and then create it with fetch enabled again.
2. The other delayed jobs (non issue tracker related) work and produce output in the logfile (dito on my other OBS instance)

I have now deactivated all trackers there except one. Lets see what happens,

**Comment by mmohring** [comment]
If you delete the last issue tracker, you get:

Server returned an error: HTTP Error 500: Internal Server Error

Request: MYAPIURL/issue_trackers/bnc
Headers:
status: 500 Internal Server Error
x-request-id: ab5c31cf-97c4-4b01-a1af-ec97bfac77bb
x-powered-by: Phusion Passenger 5.0.30
connection: close
content-length: 131
server: Apache
x-runtime: 0.034612
cache-control: public
date: Tue, 18 Jul 2017 10:00:15 GMT
content-type: application/xml; charset=utf-8

$  osc -A MYAPIURL api issue_trackers 
<nil-classes/>

**Comment by mmohring** [comment]
I have now set one issue tracker, lets see what happens in the logfile.

**Comment by mmohring** [comment]
I get two errors for the fetching jobs now:

E, [2017-07-18T12:17:16.540999 #29299] ERROR -- : [29299:272.84] 2017-07-18T12:17:16+0200: [Worker(delayed_job.1010 host:api pid:29299)] Job Delayed::PerformableMethod (id=1247318) (queue=issuetracking) FAILED permanently with Delayed::DeserializationError: ActiveRecord::RecordNotFound, class: IssueTracker, primary key: 30 (Couldn't find IssueTracker with 'id'=30)
I, [2017-07-18T12:17:21.572420 #29299]  INFO -- : [29299:277.87] 2017-07-18T12:17:21+0200: [Worker(delayed_job.1010 host:api pid:29299)] Job Delayed::PerformableMethod (id=1247320) (queue=issuetracking) RUNNING
E, [2017-07-18T12:17:21.578954 #29299] ERROR -- : [29299:277.88] 2017-07-18T12:17:21+0200: [Worker(delayed_job.1010 host:api pid:29299)] Job Delayed::PerformableMethod (id=1247320) (queue=issuetracking) FAILED permanently with Delayed::DeserializationError: ActiveRecord::RecordNotFound, class: IssueTracker, primary key: 32 (Couldn't find IssueTracker with 'id'=32)

**Comment by mmohring** [comment]
So let me conclude what I did:
1. deleted *all* issue trackers. This gives an error in the last tracker.
2. added two trackers, bnc and cve.
3. restarted delayed jobs
When delayed jobs now calls the two fetching operations, they fail with above error.

**Comment by evanrolfe** [comment]
> 1. The issue trackers delayed jobs fail
> 2. You cannot delete a issue tracker
> 3. An update works in the API but still lets the delayed jobs fail
> 4. A disabled issue tracker cannot be enabled by the API

Theres a lot going on here... Although these might be related I think it would be easier for us to discuss each point in its own issue because its getting a bit hard to follow!

> F, [2017-07-16T22:56:50.751102 #19890] FATAL -- : [93c1638d-d6a6-4320-bc10-aba46d4e198e] [19890:7636.93] ActiveRecord::StatementInvalid (Mysql2::Error: Lock wait timeout exceeded; try restarting transaction: DELETE FROM package_issues WHERE package_issues.

For this error, you can try removing the `lock(true)` part of the queries on line 18 and 29. Since i can't recreate your error I have no idea if this is actually the cause but its worth a try I would say.

> E, [2017-07-18T12:17:16.540999 #29299] ERROR -- : [29299:272.84] 2017-07-18T12:17:16+0200: [Worker(delayed_job.1010 host:api pid:29299)] Job Delayed::PerformableMethod (id=1247318) (queue=issuetracking) FAILED permanently with Delayed::DeserializationError: ActiveRecord::RecordNotFound, class: IssueTracker, primary key: 30 (Couldn't find IssueTracker with 'id'=30)

This is happening because you deleted the issue trackers manually. You will need to delete the delayed jobs too.

**Comment by mmohring** [comment]
> This is happening because you deleted the issue trackers manually. You will need to delete the delayed jobs too.
How do I delete the delayed jobs, and why is that needed ?

**Comment by evanrolfe** [comment]
Because of the way DelayedJob works: it saves a delayed job to the database with a yaml serialised payload which contains an `IssueTracker` instance, but since you have deleted the issue trackers manually those serialised objects point to issue trackers which don't exist in the database, hence why you get this error:
> Delayed::DeserializationError: ActiveRecord::RecordNotFound, class: IssueTracker, primary key: 30 (Couldn't find IssueTracker with 'id'=30)

So those delayed jobs can't run but since the issue trackers dont exist anymore we can just delete them with:

```ruby
Delayed::Job.where(queue: 'issuetracking').delete_all
```

**Comment by mmohring** [comment]
Are the delayed jobs for issue tracking automatically created when you call the API to create a new tracker ?

**Comment by evanrolfe** [comment]
Yes the delayed jobs are created automatically when an issue tracker is created (through API or UI).

**Comment by mmohring** [comment]
@evanrolfe so can you give me a hint what is the solution ? Before we do that what is the problem. I thought that all what you say is done automatically by OBS. E.g. I call the API and then everything is put into the DB and the delayed jobs are automatically created or deleted. My initial problem came from the fact that:

1. I could not update the issue tracker lists enabe / disable flag
2. I could not get away the error in the delayed jobs daemon
3. How do I tell the apidelayed jobs daemon to adapt to changes of the issue tracker list (not by fixing the code but how in practise). And is this a bug or not.

Maybe I misunderstood the API, then tell me please how it is supposed to work, e.g. the two points I list.

**Comment by adrianschroeter** [comment]
On Dienstag, 18. Juli 2017, 12:22:43 CEST wrote Martin Mohring:
> So let me conclude what I did:
> 1. deleted *all* issue trackers. This gives an error in the last tracker.

why do you do that in first place? Just disabling the fetching should be enough, no?
(and most are disabled by default)

> 2. added two trackers, bnc and cve.
> 3. restarted delayed jobs
> When delayed jobs now calls the two fetching operations, they fail with above error.
> 
> 


-- 

Adrian Schroeter
email: adrian@suse.de

SUSE Linux GmbH, GF: Felix Imendörffer, Jane Smithard, Graham Norton, HRB 21284 (AG Nürnberg) 
 
Maxfeldstraße 5                         
90409 Nürnberg 
Germany

**Comment by mmohring** [comment]
> Yes the delayed jobs are created automatically when an issue tracker is created (through API or UI).

How do I do this with the API or UI ?

**Comment by mmohring** [comment]
@adrianschroeter Hi, I dont see your comment you made here. But I got an e-mail notification for it.

**Comment by mmohring** [comment]
I have attached a excerpt from my logfile with the delayed_job failure: 
[production-log.txt](https://github.com/openSUSE/open-build-service/files/1157556/production-log.txt)
Is there a method how to recreate the delayed_jobs database if I have that database borked ?

**Comment by mmohring** [comment]
I have found a solution for all my problems:

1. rcobsapidelayed stop
2. drop the 1.5 GB sized delayed_jobs DB table
3. create an empty delayed_jobs DB table
4. rcobsapidelayed start

Then the delayed jobs errors in the logfile also disappeared. And OBS is fast now.

I did not have to fiddle with the issue trackers. Does OBS not garbage collect the DB table delayed_jobs ? How do you handle that in your big b.o.o instance ?

**Comment by mmohring** [comment]
Anyway, one bug remains: updating the fetching flag in an issue tracker with `<enable-fetch>true</enable-fetch>` does not work. I will file a separate issue for that.

**Comment by evanrolfe** [comment]
> So whats the problem then ?

I dont know, we don't have this problem on our instances and we are unable to recreate it locally so its very hard for me to tell you whats going wrong with yours. 

> Is there a method how to recreate the delayed_jobs database if I have that database borked ?

Unfortunately not. The way delayed jobs are set up now is a bit of a mess so this is not really possible at the moment. We've also been having problems with the delayed jobs setup and are currently in the process of revising this part of the system.

> Does OBS not garbage collect the DB table delayed_jobs?

Yes it should delete the processed delayed jobs automatically, not sure why this wasn't happening for you. Sorry I couldn't be of more help.

> Anyway, one bug remains: updating the fetching flag in an issue tracker with <enable-fetch>true</enable-fetch> does not work. I will file a separate issue for that.

Ok sounds good, I will close this issue then. Thanks
